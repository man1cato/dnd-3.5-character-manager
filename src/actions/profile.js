import database from '../firebase/firebase'


//SET PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
});

//READ PROFILE FROM FIREBASE
export const startSetProfile = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val();
            dispatch(setProfile(id, profile));
        });
    };
};

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
});

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates));
        });
    };
};

//DELETE PROFILE FROM STORE
export const deleteProfile = () => ({
    type: 'DELETE_PROFILE'
})

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
