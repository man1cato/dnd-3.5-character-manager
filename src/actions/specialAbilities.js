import database from '../firebase/firebase'
// import getAirtableSpecialAbilities from '../utils/getAirtableSpecialAbilities';

//SET SPECIAL ABILITIES IN STORE
export const setSpecialAbilities = (specialAbilities) => ({
    type: 'SET_SPECIAL_ABILITIES',
    specialAbilities
});

//GET SPECIAL ABILITIES FROM AIRTABLE & DISPATCH TO STORE
// export const startSetSpecialAbilities = () => {
//     return (dispatch) => {
//       return getAirtableSpecialAbilities().then((specialAbilities) => {
//         dispatch(setSpecialAbilities(specialAbilities));
//       });
//     };
// };

//GET SPECIAL ABILITIES FROM FIREBASE & DISPATCH TO STORE
export const startSetSpecialAbilities = () => {
  return (dispatch) => {
    return database.ref('api/specialAbilities').once('value').then((snapshot) => {
      const specialAbilities = snapshot.val()
      dispatch(setSpecialAbilities(specialAbilities))
    })
  }
}

