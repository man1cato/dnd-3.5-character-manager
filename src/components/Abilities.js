import React from 'react';

const Abilities = ({ abilities, handleChange }) => (
	<div className="grid grid--abilities">
		<h5 className="grid__col1">Ability</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col1" key={i}>{abilities[ability].name}</div>
		)}

		<h5 className="grid__col2">Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col2" key={i}>{abilities[ability].score}</div>
		)}

		<h5 className="grid__col3">Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col3" key={i}>{abilities[ability].mod}</div>
		)}

		<h5 className="grid__col4">Temp Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<input
			className="grid__col4"
			key={i}
			id={ability}
			name="abilities"
			type="text"
			value={abilities[ability].tempScore}
			onChange={handleChange}
			/>
		)}

		<h5 className="grid__col5">Temp Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col5" key={i}>{abilities[ability].tempMod}</div>
		)}
	</div>
)

export default Abilities;
