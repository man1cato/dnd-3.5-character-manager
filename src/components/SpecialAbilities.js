import React, { useState } from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'

import SpecialAbilityModal from './SpecialAbilityModal'


export const SpecialAbilities = (props) => {
	const [selected, setSelected] = useState(undefined)	 
	const specialAbilities = _.sortBy(props.specialAbilityIds.map((id) => {
		return {
			...props.specialAbilities[id],
			id
		}
	}), ['name'])

	return (
		<div>
			{specialAbilities.map((specialAbility, i) => (
				<button
					className="button--link"
					id={specialAbility.id}
					key={`specialAbility${i}`}
					onClick={() => setSelected(specialAbility)}
				>
					{specialAbility.name}
				</button>
			))}

			<SpecialAbilityModal 
				selected={selected}
				handleCloseModal={() => setSelected(undefined)}
			/>
		</div>
	)	
}

const mapStateToProps = (state) => ({
  	specialAbilities: state.specialAbilities
})

export default connect(mapStateToProps)(SpecialAbilities)