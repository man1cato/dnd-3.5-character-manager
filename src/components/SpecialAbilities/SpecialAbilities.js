import React, { useState } from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'

import SpecialAbilityModal from '../Modals/SpecialAbilityModal'


export const SpecialAbilities = (props) => {
	const [clickedAbility, setClickedAbility] = useState(null)	 
	const specialAbilities = _.sortBy(props.specialAbilityIds.map((id) => {
		return {
			...props.specialAbilities[id],
			id
		}
	}), ['name'])

	return (
		<div>
			{specialAbilities.map((specialAbility, i) => (
				<button
					className="button--link"
					id={specialAbility.id}
					key={`specialAbility${i}`}
					type='button'
					onClick={() => setClickedAbility(specialAbility)}
				>
					{specialAbility.name}
				</button>
			))}

			<SpecialAbilityModal 
				clickedAbility={clickedAbility}
				handleCloseModal={() => setClickedAbility(null)}
			/>
		</div>
	)	
}

const mapStateToProps = (state) => ({
  	specialAbilities: state.specialAbilities
})

export default connect(mapStateToProps)(SpecialAbilities)