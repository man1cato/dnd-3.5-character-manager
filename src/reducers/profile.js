export default (state = {}, action) => {
    switch (action.type) {
        case 'SET_PROFILE':
            return {
                id: action.id,
                ...action.fields
            };
        case 'EDIT_PROFILE':
            return {
                ...state,
                ...action.updates
            };
        default:
            return state;
    }
};
