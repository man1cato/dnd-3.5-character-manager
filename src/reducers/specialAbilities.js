export default (state = {}, action) => {
    switch (action.type) {
        case 'SET_SPECIALABILITIES':
            return [
                ...action.specialAbilities
            ];
        default:
            return state;
    }
};
