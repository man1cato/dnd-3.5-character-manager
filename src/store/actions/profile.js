import database from '../../firebase/firebase'
import { startGetProfiles } from './profiles'

//SET SELECTED PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
})

//READ PROFILE FROM FIREBASE
export const startSetProfile = (uid, id) => {
    return (dispatch) => {
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val()
            dispatch(setProfile(id, profile))
        })
    }
}

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
            dispatch(startGetProfiles(uid))
            localStorage.setItem('selectedCharacterId', ref.key)
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
})

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates))
        })
    }
}

//REMOVE PROFILE FROM STORE
export const removeProfile = () => ({
    type: 'REMOVE_PROFILE'
})

//REMOVE PROFILE FROM FIREBASE
// export const startRemoveProfile = () => {
//     return (dispatch) => {
//         return database.ref(`users/${uid}/profiles/${id}`).remove().then(() => {
//             dispatch(removeProfile())
//         })
//     }
// }

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
