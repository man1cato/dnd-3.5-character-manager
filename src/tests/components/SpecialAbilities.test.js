import React from 'react';
import {shallow} from 'enzyme';

import { SpecialAbilities } from '../../components/SpecialAbilities';
import profile from '../fixtures/profile'
import specialAbilities from '../fixtures/specialAbilities'


const props = {
  specialAbilityIds: profile.fields.specialAbilities,
  specialAbilities
}

test('should render special abilities with profile data', () => {
  const wrapper = shallow(<SpecialAbilities {...props} />);
  expect(wrapper).toMatchSnapshot();
})
