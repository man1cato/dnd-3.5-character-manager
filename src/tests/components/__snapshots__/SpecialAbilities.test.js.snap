// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render special abilities with profile data 1`] = `
<div>
  <button
    className="button--link"
    id="rect54ikkC9SDLJ1x"
    key="specialAbility0"
    onClick={[Function]}
    type="button"
  >
    A Thousand Faces
  </button>
  <SpecialAbilityModal
    handleCloseModal={[Function]}
  />
</div>
`;
